File: apps/desktop/src/lib/llmClient.ts
import { LLMProvider, LLMResponse } from '@core/types';

/**
 * URL do Cloudflare Worker proxy (resolve CORS de providers como NVIDIA NIM).
 * Defina a variavel de ambiente VITE_PROXY_URL no Cloudflare Pages
 * apontando para o worker: https://llm-proxy.<subdominio>.workers.dev
 *
 * Se nao definida, tenta chamada direta (funciona para OpenAI, Groq, Together que tem CORS).
 */
const PROXY_URL = (import.meta as any).env?.VITE_PROXY_URL as string | undefined;

/**
 * Providers que tem CORS aberto — chamada direta sem proxy.
 */
const DIRECT_ORIGINS = [
  'api.openai.com',
  'api.groq.com',
  'api.together.xyz',
  'api.mistral.ai',
  'openrouter.ai',
];

function needsProxy(baseUrl: string): boolean {
  if (!PROXY_URL) return false;
  try {
    const host = new URL(baseUrl).hostname;
    return !DIRECT_ORIGINS.some(o => host.includes(o));
  } catch {
    return false;
  }
}

export async function callLLM(
  provider: LLMProvider,
  prompt: string,
  temperature: number = 0.7
): Promise<LLMResponse> {
  if (!provider?.apiKey) {
    throw new Error('Nenhuma API Key configurada. Vá em Configurações e adicione um provedor.');
  }

  const baseUrl = (provider.baseUrl || 'https://api.openai.com/v1').replace(/\/$/, '');
  const endpoint = `${baseUrl}/chat/completions`;
  const useProxy = needsProxy(baseUrl);

  const fetchUrl     = useProxy ? PROXY_URL! : endpoint;
  const extraHeaders = useProxy ? { 'X-Target-URL': endpoint } : {};

  const res = await fetch(fetchUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${provider.apiKey}`,
      ...extraHeaders,
    },
    body: JSON.stringify({
      model: provider.model || 'gpt-4o-mini',
      messages: [{ role: 'user', content: prompt }],
      temperature,
      max_tokens: 4096,
    }),
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Erro do provedor (${res.status}): ${err}`);
  }

  const data = await res.json();
  return {
    content: data.choices?.[0]?.message?.content ?? '',
    tokens_used: data.usage?.total_tokens ?? 0,
  };
}

File: apps/desktop/src-tauri/src/llm_client.rs
use serde::{Deserialize, Serialize};
use reqwest::Client;
use std::time::Duration;

#[derive(Debug, Serialize, Deserialize)]
pub struct LLMResponse {
    pub content: String,
    pub tokens_used: u32,
}

#[derive(Serialize)]
struct GenericChatRequest {
    model: String,
    messages: Vec<Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f32>,
}

#[derive(Serialize)]
struct Message {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct GenericChatResponse {
    choices: Vec<Choice>,
    usage: Option<Usage>,
}

#[derive(Deserialize)]
struct Choice {
    message: ResponseMessage,
}

#[derive(Deserialize)]
struct ResponseMessage {
    content: String,
}

#[derive(Deserialize)]
struct Usage {
    total_tokens: u32,
}

#[tauri::command]
pub async fn call_llm(
    base_url: String,
    api_key: String,
    model: String,
    prompt: String,
    temperature: Option<f32>,
) -> Result<LLMResponse, String> {
    let client = Client::builder()
        .timeout(Duration::from_secs(120))
        .build()
        .map_err(|e| e.to_string())?;

    // Normaliza a URL para garantir que termine com /v1/chat/completions se necessário
    let url = if base_url.ends_with("/chat/completions") {
        base_url
    } else if base_url.ends_with("/") {
        format!("{}chat/completions", base_url)
    } else {
        format!("{}/chat/completions", base_url)
    };

    let body = GenericChatRequest {
        model,
        messages: vec![Message {
            role: "user".to_string(),
            content: prompt,
        }],
        temperature,
    };

    let res = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| e.to_string())?;

    if !res.status().is_success() {
        let error_text = res.text().await.unwrap_or_else(|_| "Erro desconhecido".to_string());
        return Err(format!("AI Provider Error ({}): {}", url, error_text));
    }

    let data: GenericChatResponse = res.json().await.map_err(|e| e.to_string())?;
    
    let content = data.choices.get(0)
        .map(|c| c.message.content.clone())
        .ok_or("Resposta da IA vazia")?;

    let tokens = data.usage.map(|u| u.total_tokens).unwrap_or(0);

    Ok(LLMResponse {
        content,
        tokens_used: tokens,
    })
}
